scale/job/models.py

```diff
@@ -212,55 +206,145 @@ def get_job_updates(self, started=None, ended=None, status=None, job_type_ids=No
             order = ['last_status_change']
         return self.get_jobs(started, ended, status, job_type_ids, job_type_names, job_type_categories, order)
 
-    @transaction.atomic
-    def queue_job(self, job, data, when):
-        """Puts the given job into the QUEUED state with the given arguments. The data should be set to None if this is
-        not the first time the job has been queued. The given job model must have already been saved in the database (it
-        must have an ID), it must have its related job type and job type revision fields, and the caller must have
-        obtained a lock on the job model using select_for_update(). The changes to the job will be saved in the database
-        in an atomic transaction.
-
-        :param job: The job to put on the queue
-        :type job: :class:`job.models.Job`
-        :param data: JSON description defining the job data to run on
-        :type data: dict
-        :param when: The time that the job was queued
+    def get_locked_job(self, job_id):
+        """Gets the job model with the given ID with a model lock obtained and related job_type and job_type_rev models
+
+        :param job_id: The job ID
+        :type job_id: int
+        :returns: The job model
+        :rtype: :class:`job.models.Job`
+        """
+
+        return self.get_locked_jobs([job_id])[0]
+
+    def get_locked_jobs(self, job_ids):
+        """Gets the job models for the given IDs with model locks obtained and related job_type and job_type_rev models
+
+        :param job_ids: The job IDs
+        :type job_ids: [int]
+        :returns: The job models
+        :rtype: [:class:`job.models.Job`]
+        """
+
+        self.lock_jobs(job_ids)
+
+        return list(self.select_related('job_type', 'job_type_rev').filter(id__in=job_ids).iterator())
+
+    def increment_max_tries(self, jobs):
+        """Increments the max_tries of the given jobs to be one greater than their current number of executions. The
+        caller must have obtained model locks on the job models.
+
+        :param jobs: The jobs to update
+        :type jobs: [:class:`job.models.Job`]
+        """
+
+        modified = timezone.now()
+
+        # Update job models in memory and collect job IDs
+        job_ids = set()
+        for job in jobs:
+            job_ids.add(job.id)
+            job.max_tries = job.num_exes + 1
+            job.last_modified = modified
+
+        # Update job models in database with single query
+        self.filter(id__in=job_ids).update(max_tries=models.F('num_exes') + 1, last_modified=modified)
+
+    def lock_jobs(self, job_ids):
+        """Obtains model locks on the job models with the given IDs (in ID order to prevent deadlocks)
+
+        :param job_ids: The IDs of the jobs to lock
+        :type job_ids: [int]
+        """
+
+        list(self.select_for_update().filter(id__in=job_ids).order_by('id').iterator())
```

Comment:
Could this return a query set? When it gets called by other methods like get_locked_jobs() it could then be chained together into a single query, but once it gets wrapped as a list it forces Django to evaluate the query.